File: frontend/lib/features/seals/seal_calculations.dart
import 'seal_form_loader.dart';

/// Plocha z rozměrů v mm → m²
double areaFromMm(int lengthMm, int widthMm) =>
    (lengthMm * widthMm) / 1000000.0;

/// VZT: běžné metry
double vztLinearMeters(int lengthMm, int widthMm) =>
    ((2 * lengthMm + 2 * widthMm) * 2) / 1000.0;

/// Kruh: plocha z průměru v mm → m²
double circleAreaFromDiameterMm(int diameterMm) {
  final r = diameterMm / 2.0;
  return (3.141592653589793 * r * r) / 1000000.0;
}

/// Vytáhne průměr (Ø) z rozměru typu "Ø50" nebo "Ø20-100".
int? diameterFromDimension(String? dim) {
  if (dim == null) return null;
  final n = dim.replaceAll(RegExp(r'\s+'), '').toLowerCase();
  final range = RegExp(r'ø(\d+)-(\d+)').firstMatch(n);
  if (range != null) {
    return ((int.parse(range.group(1)!) + int.parse(range.group(2)!)) / 2).round();
  }
  final single = RegExp(r'ø(\d+)').firstMatch(n);
  if (single != null) return int.parse(single.group(1)!);
  return null;
}

class SealCalculationResult {
  const SealCalculationResult({
    this.openingAreaM2,
    this.deductionAreaM2,
    this.netAreaM2,
    this.linearMeters,
    required this.billableQuantity,
    required this.unit,
    this.netAreaWasNegative = false,
  });

  final double? openingAreaM2;
  final double? deductionAreaM2;
  final double? netAreaM2;
  final double? linearMeters;
  final double billableQuantity;
  final String unit;
  final bool netAreaWasNegative;
}

int? parseMmText(String? text) {
  if (text == null || text.trim().isEmpty) return null;
  final n = int.tryParse(text.trim());
  if (n == null || n <= 0) return null;
  return n;
}

SealCalculationResult computeSealEntryPreview({
  required String entryType,
  required int quantityKus,
  int? openingLengthMm,
  int? openingWidthMm,
  int? itemLengthMm,
  int? itemWidthMm,
  required List<SealEntryDraftData> allEntries,
  required int entryIndex,
}) {
  if (entryType == 'VZT' && itemLengthMm != null && itemWidthMm != null) {
    final mb = vztLinearMeters(itemLengthMm, itemWidthMm);
    return SealCalculationResult(
      linearMeters: mb,
      billableQuantity: mb,
      unit: 'mb',
    );
  }

  if (entryType == 'PROSTUP') {
    double? gross;
    if (itemLengthMm != null && itemWidthMm != null) {
      gross = areaFromMm(itemLengthMm, itemWidthMm);
    } else if (openingLengthMm != null && openingWidthMm != null) {
      gross = areaFromMm(openingLengthMm, openingWidthMm);
    }

    if (gross == null) {
      return SealCalculationResult(
        billableQuantity: quantityKus.toDouble(),
        unit: 'kus',
      );
    }

    final hasOpening =
        openingLengthMm != null && openingWidthMm != null;
    final openingArea = hasOpening
        ? areaFromMm(openingLengthMm, openingWidthMm)
        : gross;

    double deduction = 0;
    for (var i = 0; i < allEntries.length; i++) {
      if (i == entryIndex) continue;
      final other = allEntries[i];
      final oL = parseMmText(other.itemLengthMmText);
      final oW = parseMmText(other.itemWidthMmText);
      if (oL != null && oW != null) {
        // Obdélník: exaktní plocha bez příplatku (Task 5)
        deduction += areaFromMm(oL, oW);
      } else {
        // Kruh: pokud má rozměr průměr Ø
        final d = diameterFromDimension(other.dimension);
        if (d != null && d > 0) {
          deduction += circleAreaFromDiameterMm(d);
        }
      }
    }

    var net = gross;
    var wasNegative = false;
    if (hasOpening && deduction > 0) {
      final raw = openingArea - deduction;
      if (raw < 0) {
        net = 0;
        wasNegative = true;
      } else {
        net = raw;
      }
    }

    return SealCalculationResult(
      openingAreaM2: hasOpening ? openingArea : gross,
      deductionAreaM2: deduction > 0 ? deduction : null,
      netAreaM2: hasOpening && deduction > 0 ? net : null,
      billableQuantity: net,
      unit: 'm2',
      netAreaWasNegative: wasNegative,
    );
  }

  return SealCalculationResult(
    billableQuantity: quantityKus.toDouble(),
    unit: 'kus',
  );
}

String formatArea(double value) => value.toStringAsFixed(3);
String formatMb(double value) => value.toStringAsFixed(1);

String unitLabel(String unit) {
  switch (unit) {
    case 'm2':
      return 'm²';
    case 'mb':
      return 'mb';
    default:
      return 'ks';
  }
}

File: backend/src/services/seal-calculations.ts
/** Chybová hláška při přeodečtení (odečet > celková plocha prostupu). */
export const OVER_DEDUCTION_MESSAGE =
  'Odečtená plocha je větší než celková plocha prostupu.';

/** Obdélník: plocha z rozměrů v mm → m² */
export function areaFromMm(lengthMm: number, widthMm: number): number {
  return (lengthMm * widthMm) / 1_000_000;
}

/** Kruh: plocha z průměru v mm → m² */
export function circleAreaFromDiameterMm(diameterMm: number): number {
  const r = diameterMm / 2;
  return (Math.PI * r * r) / 1_000_000;
}

/** VZT: běžné metry z D a Š v mm */
export function vztLinearMeters(lengthMm: number, widthMm: number): number {
  return ((2 * lengthMm + 2 * widthMm) * 2) / 1000;
}

/**
 * Plocha procházející instalace v m² podle tvaru (Task 5):
 *  - obdélník: šířka × výška / 1e6
 *  - kruh: π × (Ø/2)² / 1e6
 *  - ruční: zadaná hodnota v m²
 */
export type Deduction =
  | { kind: 'rect'; widthMm: number; heightMm: number }
  | { kind: 'circle'; diameterMm: number }
  | { kind: 'manual'; areaM2: number };

export function deductionArea(d: Deduction): number {
  switch (d.kind) {
    case 'rect':
      return areaFromMm(d.widthMm, d.heightMm);
    case 'circle':
      return circleAreaFromDiameterMm(d.diameterMm);
    case 'manual':
      return Math.max(0, d.areaM2);
  }
}

/**
 * Čistá účtovaná plocha prostupu = celková plocha otvoru − součet ploch instalací.
 * Nikdy není záporná; pokud odečet přesáhne celkovou plochu, vrátí wasNegative=true.
 */
export function netOpeningArea(
  openingAreaM2: number,
  deductionAreasM2: number[],
): { netAreaM2: number; wasNegative: boolean } {
  const totalDeduction = deductionAreasM2.reduce((sum, a) => sum + a, 0);
  const raw = openingAreaM2 - totalDeduction;
  if (raw < 0) {
    return { netAreaM2: 0, wasNegative: true };
  }
  return { netAreaM2: raw, wasNegative: false };
}

export type SealOpening = {
  openingLengthMm: number | null;
  openingWidthMm: number | null;
};

export type EntryDimensions = {
  entryType: string;
  itemLengthMm: number | null;
  itemWidthMm: number | null;
  dimension?: string | null;
};

/** Vytáhne průměr (Ø) z textového rozměru typu "Ø50" nebo "Ø20-100" (průměr = průměr středu). */
function diameterFromDimension(dim: string | null | undefined): number | null {
  if (!dim) return null;
  const n = dim.replace(/\s+/g, '').toLowerCase();
  const range = n.match(/ø(\d+)-(\d+)/);
  if (range) return Math.round((parseInt(range[1], 10) + parseInt(range[2], 10)) / 2);
  const single = n.match(/ø(\d+)/);
  if (single) return parseInt(single[1], 10);
  return null;
}

export type ComputedEntryValues = {
  calculatedAreaM2: number | null;
  calculatedLinearMeters: number | null;
  calculatedNetAreaM2: number | null;
  billableQuantity: number;
  unit: 'kus' | 'm2' | 'mb';
  netAreaWasNegative: boolean;
};

function hasMmPair(length: number | null, width: number | null): length is number {
  return length != null && width != null && length > 0 && width > 0;
}

/**
 * Plochy procházejících instalací (exaktně, Task 5):
 *  - má-li instalace obdélníkové rozměry (item D×Š) → obdélník
 *  - jinak má-li v rozměru průměr Ø → kruh
 *  - ostatní (bez rozměru) se neodečítají
 */
export function sumDeductionAreas(entries: EntryDimensions[]): number[] {
  const areas: number[] = [];
  for (const e of entries) {
    if (hasMmPair(e.itemLengthMm, e.itemWidthMm)) {
      areas.push(areaFromMm(e.itemLengthMm!, e.itemWidthMm!));
      continue;
    }
    const d = diameterFromDimension(e.dimension);
    if (d != null && d > 0) {
      areas.push(circleAreaFromDiameterMm(d));
    }
  }
  return areas;
}

export function computeEntryValues(
  entry: EntryDimensions & { quantity: number },
  sealOpening: SealOpening,
  allEntries: EntryDimensions[],
  entryIndex: number,
): ComputedEntryValues {
  const defaultResult: ComputedEntryValues = {
    calculatedAreaM2: null,
    calculatedLinearMeters: null,
    calculatedNetAreaM2: null,
    billableQuantity: entry.quantity,
    unit: 'kus',
    netAreaWasNegative: false,
  };

  if (entry.entryType === 'VZT' && hasMmPair(entry.itemLengthMm, entry.itemWidthMm)) {
    const mb = vztLinearMeters(entry.itemLengthMm!, entry.itemWidthMm!);
    return {
      ...defaultResult,
      calculatedLinearMeters: mb,
      billableQuantity: mb,
      unit: 'mb',
    };
  }

  if (entry.entryType === 'PROSTUP') {
    let grossArea: number | null = null;

    if (hasMmPair(entry.itemLengthMm, entry.itemWidthMm)) {
      grossArea = areaFromMm(entry.itemLengthMm!, entry.itemWidthMm!);
    } else if (
      hasMmPair(sealOpening.openingLengthMm, sealOpening.openingWidthMm)
    ) {
      grossArea = areaFromMm(sealOpening.openingLengthMm!, sealOpening.openingWidthMm!);
    }

    if (grossArea == null) {
      return defaultResult;
    }

    const openingSet = hasMmPair(sealOpening.openingLengthMm, sealOpening.openingWidthMm);
    const openingArea = openingSet
      ? areaFromMm(sealOpening.openingLengthMm!, sealOpening.openingWidthMm!)
      : grossArea;

    const otherEntries = allEntries.filter((_, i) => i !== entryIndex);
    const deductions = sumDeductionAreas(otherEntries);

    let netAreaM2 = grossArea;
    let wasNegative = false;

    if (openingSet && deductions.length > 0) {
      const net = netOpeningArea(openingArea, deductions);
      netAreaM2 = net.netAreaM2;
      wasNegative = net.wasNegative;
    }

    return {
      calculatedAreaM2: grossArea,
      calculatedLinearMeters: null,
      calculatedNetAreaM2: openingSet && deductions.length > 0 ? netAreaM2 : null,
      billableQuantity: netAreaM2,
      unit: 'm2',
      netAreaWasNegative: wasNegative,
    };
  }

  return defaultResult;
}
